hermes/crates/cardano-chain-follower/src/metadata/cip509/mod.rs

```diff
@@ -0,0 +1,954 @@
+#[derive(FromRepr, Debug, PartialEq)]
+#[repr(u8)]
+pub(crate) enum Cip509Int {
+    /// Purpose.
+    Purpose = 0,
+    /// Transaction inputs hash.
+    TxInputsHash = 1,
+    /// Previous transaction ID.
+    PreviousTxId = 2,
+    /// Validation signature.
+    ValidationSignature = 99,
+}
+
+impl Decode<'_, ()> for Cip509 {
+    fn decode(d: &mut Decoder, ctx: &mut ()) -> Result<Self, decode::Error> {
+        let map_len = decode_map_len(d, "CIP509")?;
+        let mut cip509_metadatum = Cip509::default();
+        for _ in 0..map_len {
+            // Use probe to peak
+            let key = d.probe().u8()?;
+            if let Some(key) = Cip509Int::from_repr(key) {
+                // Consuming the int
+                decode_u8(d, "CIP509")?;
+                match key {
+                    Cip509Int::Purpose => {
+                        cip509_metadatum.purpose = decode_bytes(d, "CIP509 purpose")?
+                            .try_into()
+                            .map_err(|_| decode::Error::message("Invalid data size of Purpose"))?;
+                    },
+                    Cip509Int::TxInputsHash => {
+                        cip509_metadatum.txn_inputs_hash =
+                            decode_bytes(d, "CIP509 txn inputs hash")?
+                                .try_into()
+                                .map_err(|_| {
+                                    decode::Error::message("Invalid data size of TxInputsHash")
+                                })?;
+                    },
+                    Cip509Int::PreviousTxId => {
+                        cip509_metadatum.prv_tx_id = Some(
+                            decode_bytes(d, "CIP509 previous tx ID")?
+                                .try_into()
+                                .map_err(|_| {
+                                    decode::Error::message("Invalid data size of PreviousTxId")
+                                })?,
+                        );
+                    },
+                    Cip509Int::ValidationSignature => {
+                        let validation_signature = decode_bytes(d, "CIP509 validation signature")?;
+                        if validation_signature.is_empty() || validation_signature.len() > 64 {
+                            return Err(decode::Error::message(
+                                "Invalid data size of ValidationSignature",
+                            ));
+                        }
+                        cip509_metadatum.validation_signature = validation_signature;
+                    },
+                }
+            } else {
+                // Handle the x509 chunks 10 11 12
+                let x509_chunks = X509Chunks::decode(d, ctx)?;
+                cip509_metadatum.x509_chunks = x509_chunks;
+            }
+        }
+        Ok(cip509_metadatum)
+    }
+}
+
+#[allow(clippy::module_name_repetitions)]
+impl Cip509 {
+    /// Decode and validate CIP509 metadata.
+    pub(crate) fn decode_and_validate(
```

Comment:
Dont we need to return some decoded data here ?
If we are not returning anything name of this function sounds confusing.

Reply:
Nope, the necessary data is already kept in the struct itself and all the possible error are handle in the `validation_report` where the user can see the entire error report